### ISSUE TITLE
Processing stops because sessions of the DistributedLogStreamClient are closed

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#3370]
**Describe the bug**

Lately we observed in our long running test suite that the processing of one partition (eventually most of the partitions) stop processing.

![processing-stopped](https://user-images.githubusercontent.com/2758593/68862559-f5424c00-06ed-11ea-88ae-35f1a7df96d3.png)

We saw no broker restart nor leader change at this time.
The only thing we saw at this point in time was the reconfiguration of the DistributedLogPrimitive.
That is the reason why we assumed the session might be closed and the append was not happening afterwards.

If we take a look at the Heartbeat rate we see that at some point only the real heartbeat requests are send, no longer any append requests.

![heartbeats](https://user-images.githubusercontent.com/2758593/68862733-44887c80-06ee-11ea-9cf6-61a1cb83364c.png)

In the log we saw log statements like this:

```
I 2019-11-14T13:29:40.145523760Z zeebe-1 2019-11-14 13:29:40.144 [] [raft-partition-group-raft-atomix-13] ERROR io.zeebe.logstreams - Failed to append block with last event position 1099516985048, retry.
  zeebe-1
I 2019-11-14T13:29:40.145546797Z zeebe-1 java.util.concurrent.CompletionException: io.atomix.primitive.PrimitiveException$ClosedSession: session closed
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
	at io.atomix.utils.concurrent.Futures.lambda$null$3(Futures.java:159) ~[atomix-utils-3.2.0-20191112.114214-59.jar:?]
	at io.atomix.utils.concurrent.ThreadPoolContext.lambda$new$0(ThreadPoolContext.java:81) ~[atomix-utils-3.2.0-20191112.114214-59.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: io.atomix.primitive.PrimitiveException$ClosedSession: session closed
	at io.atomix.protocols.raft.session.impl.RaftSessionInvoker.close(RaftSessionInvoker.java:217) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.DefaultRaftSessionClient.lambda$null$0(DefaultRaftSessionClient.java:236) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionState.lambda$setState$0(RaftSessionState.java:221) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(Unknown Source) ~[?:?]
	at java.util.concurrent.CopyOnWriteArraySet.forEach(Unknown Source) ~[?:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionState.setState(RaftSessionState.java:221) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionInvoker$CommandAttempt.accept(RaftSessionInvoker.java:388) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionInvoker$CommandAttempt.accept(RaftSessionInvoker.java:327) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.complete(Unknown Source) ~[?:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionConnection.handleResponse(RaftSessionConnection.java:282) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionConnection.lambda$sendRequest$7(RaftSessionConnection.java:238) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(Unknown Source) ~[?:?]
	... 7 more
  zeebe-1
I 2019-11-14T13:29:40.146362510Z zeebe-1 2019-11-14 13:29:40.145 [] [raft-partition-group-raft-atomix-13] ERROR io.zeebe.logstreams - Failed to append block with last event position 1099516999056, retry.
  zeebe-1
I 2019-11-14T13:29:40.146404536Z zeebe-1 java.util.concurrent.CompletionException: io.atomix.primitive.PrimitiveException$ClosedSession: session closed
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
	at io.atomix.utils.concurrent.Futures.lambda$null$3(Futures.java:159) ~[atomix-utils-3.2.0-20191112.114214-59.jar:?]
	at io.atomix.utils.concurrent.ThreadPoolContext.lambda$new$0(ThreadPoolContext.java:81) ~[atomix-utils-3.2.0-20191112.114214-59.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: io.atomix.primitive.PrimitiveException$ClosedSession: session closed
	at io.atomix.protocols.raft.session.impl.RaftSessionInvoker.close(RaftSessionInvoker.java:217) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.DefaultRaftSessionClient.lambda$null$0(DefaultRaftSessionClient.java:236) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionState.lambda$setState$0(RaftSessionState.java:221) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(Unknown Source) ~[?:?]
	at java.util.concurrent.CopyOnWriteArraySet.forEach(Unknown Source) ~[?:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionState.setState(RaftSessionState.java:221) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionInvoker$CommandAttempt.accept(RaftSessionInvoker.java:388) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionInvoker$CommandAttempt.accept(RaftSessionInvoker.java:327) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.complete(Unknown Source) ~[?:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionConnection.handleResponse(RaftSessionConnection.java:282) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at io.atomix.protocols.raft.session.impl.RaftSessionConnection.lambda$sendRequest$7(RaftSessionConnection.java:238) ~[atomix-raft-3.2.0-20191112.114310-59.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(Unknown Source) ~[?:?]
	... 7 more
  zeebe-1
I 2019-11-14T13:29:40.147062892Z zeebe-1 2019-11-14 13:29:40.146 [] [raft-partition-group-raft-atomix-13] ERROR io.zeebe.logstreams - Failed to append block with last event position 1099516999352, retry.
  zeebe-1
```

As you can see in the following code from the `LogStorageAppender`:

```java
  private void appendToPrimitive(byte[] bytesToAppend, long lastEventPosition) {
    actor.submit(
        () -> {
          distributedLog
              .asyncAppend(bytesToAppend, lastEventPosition)
              .whenComplete(
                  (appendPosition, error) -> {
                    if (error != null) {
                      LOG.error(
                          "Failed to append block with last event position {}, retry.",
                          lastEventPosition,
                          error);
                      appendToPrimitive(bytesToAppend, lastEventPosition);
                    } else {
                      actor.run(
                          () -> {
                            appendEntryLimiter.onCommit(lastEventPosition);
                            currentInFlightBytes -= bytesToAppend.length;
                            actor.run(this::peekBlock);
                          });
                    }
                  });
        });
  }
```

The append is retried on error, but `actor.submit` will throw an exception since it is not called in the `ActorThread` it self. This means the retry will fail.

But even if the retry will succeed we would have the problem that it might happen that events are then appended out of order. 

**Example**:

We append `1, 2, 3, 4` and the session close exception will happen for 2 and 3. Then it can happen that we have in the end `1, 4, 2, 3` as appended entries. This might be detected by the `DistributedLogPrimitive`, such that it does not append `2` and `3` to the log storage.

**Comment by Zelldon** [comment]
It seems to happen deterministically after 30 min.

**Comment by Zelldon** [comment]
### Observation

As described above the problem lies in reconfiguration of the DistributedLogStream and the closed sessions. In the RaftServiceManager, after a snapshot is taken, the newest valid snapshot is installed. This means that the RaftServiceContext is recreated, all sessions are removed from the old context and the Primitive is reconfigured. 

Yesterday evening I removed the step, where it removed the existing sessions [here](https://github.com/zeebe-io/atomix/blob/zeebe/protocols/raft/src/main/java/io/atomix/protocols/raft/impl/RaftServiceManager.java#L320), because my assumption was it can still reuse them. I run the benchmark over the night and as first it looked better then before.

_Processing:_
![processoing](https://user-images.githubusercontent.com/2758593/68922054-f53c5d80-077a-11ea-935d-567a7dd03302.png)

I mean it run longer then before and made progress.
But as you can see we have periodically an drop to zero. This happens every 15 min's. My current assumption is that this is related to our snapshot interval and at this time snapshots are replicated. It might make sense to randomize the snapshot interval a bit, such that not each partition takes snapshots at the same time.

_Backpressure:_
![backpressure](https://user-images.githubusercontent.com/2758593/68922055-f5d4f400-077a-11ea-9827-79380c3c48bc.png)


_Atomix metrics:_
![atomix-metrics](https://user-images.githubusercontent.com/2758593/68922056-f5d4f400-077a-11ea-95c2-cd86da567273.png)

As we can see from the backpressure at `0:43`, all appends are dropped for partition 6. We see the same in the processing graph that for partition 6 the processing drops to zero and never comes back. 

After taking a look in the log we see the same issue, as described above, which means we just deferred the issue. 

```
I 2019-11-14T23:39:28.381883768Z zeebe-0 2019-11-14 23:39:28.381 [] [raft-server-0-raft-atomix-partition-6-state] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService-6 - Configuring distributed-log on node 0 with logName raft-atomix-partition-6
  zeebe-0
I 2019-11-14T23:39:28.381927633Z zeebe-0 2019-11-14 23:39:28.381 [] [raft-server-0-raft-atomix-partition-6-state] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService-6 - Configured with LogStream raft-atomix-partition-6 and last appended event at position 19743967985120
  zeebe-0
I 2019-11-14T23:39:28.381938276Z zeebe-0 2019-11-14 23:39:28.381 [] [raft-server-0-raft-atomix-partition-6-state] ERROR io.atomix.protocols.raft.impl.RaftServiceManager - RaftServer{raft-atomix-partition-6} - Not removing sessions \-.-/
  zeebe-0
I 2019-11-14T23:39:28.385315381Z zeebe-0 2019-11-14 23:39:28.385 [] [raft-server-0-raft-atomix-partition-6-state] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService-6 - Restored local log to position 19743967985120
  zeebe-0

...

I 2019-11-14T23:43:01.935834319Z zeebe-0 2019-11-14 23:43:01.935 [] [raft-partition-group-raft-atomix-6] ERROR io.zeebe.logstreams - Failed to append block with last event position 19971597981064, retry.
  zeebe-0
I 2019-11-14T23:43:01.935869008Z zeebe-0 2019-11-14 23:43:01.935 [] [raft-partition-group-raft-atomix-6] ERROR io.zeebe.logstreams - Failed to append block with last event position 19971597981368, retry.
  zeebe-0
I 2019-11-14T23:43:01.935877017Z zeebe-0 2019-11-14 23:43:01.935 [] [raft-partition-group-raft-atomix-6] ERROR io.zeebe.logstreams - Failed to append block with last event position 19971597981672, retry.
  zeebe-0

```


### Solution

It might make sense to fix the issue in atomix. But this might not so easy and we want to get rid of the primitive anyway so it might make no sense to do that. We could avoid recreation the RaftServiceContext etc but it can still happen that sessions are expired or closed which then cause the same issues. 

_A possible solution in the Broker would be the following:_

We add an parameter on the primitive method `append`, which is called `index`. This index is increment by one on each append, by the log appender. 

The primitive always stores the last appended index. When an append entry comes with an index bigger then `lastIndex+1`, then we know that we missed some events. We will store this event until we get the previous ones. When we get the event before, we also check our queue whether we have still pending events. If this is the case we append them and update the last index, such that we always append the events in the right order. If we never get the right events before we can't continue with appending, which is the same behavior as now. Currently I expected that not so much events are coming in the wrong order and that this happens not so often. :crossed_fingers: 
This is more a quick fix until we get the new log stream.

Startup might be interesting. The primitive snapshots the lastIndex and provides an query method to access that. The log append will on startup query the last index, such that the index always increases correctly and this should also work on fail over.

**Comment by Zelldon** [comment]
I was wondering why this not happend before. I finally know why. We never handeled errors on `asyncAppend` before https://github.com/zeebe-io/zeebe/blob/0.21.0/logstreams/src/main/java/io/zeebe/logstreams/impl/LogStorageAppender.java#L79 🤦‍♂️

**Comment by Zelldon** [comment]
I run several benchmarks over the weekend. The following was with a constant load and enough workers to complete the instances:

![no-leader-install-steady-load](https://user-images.githubusercontent.com/2758593/69030675-93792f00-09d8-11ea-88bf-e5092aea731b.png)

This contains https://github.com/zeebe-io/atomix/pull/69 only. 
As you can see it died after 20 hours, but this seems to be related to https://github.com/zeebe-io/zeebe/issues/3181 .

If we reduce the workers, which means not all workflow instances are completed this increases the snapshot size and also causes also other problems it seems.

![no-leader-install-less-worker](https://user-images.githubusercontent.com/2758593/69032275-629af900-09dc-11ea-9b74-7d1ebb2df0b8.png)

As you can see with the fix of https://github.com/zeebe-io/atomix/pull/69 it is still possible, but less likely that the processing fails. Furthermore due to snapshot taking and replication we have every 15 min an drop to zero. (Related https://github.com/zeebe-io/zeebe/issues/3376)

**Comment by Zelldon** [comment]
I discussed with @deepthidevaki the solution above. We concluded that it is not ideal to defer/store the events in the primitive. This can lead to several other problems.

Problematic scenario with stored deferred events:

Say we send event 2, 3, 4. If 3 failed because of reasons. Then with the solution above we deferred the event 4. This happens and the leader as well on the follower. If now a leader changes happens, the primitive does not restart but the log appender is started. The append will query the last successful append position, which is 2. It will send the next event with idx 3 which is not the same as the 3 before. The primitive will apply 4 afterwards and increment the idx. The next event will be ignored by the primitive because it has the same idx, this results in inconsistent log and indices in appender and primitive. This problem also happens if we store the deferred events in the backup - this cause also much more problems. 

This is the reason we decided to just return an negative result and retry on the append if the result is negative. Since we retry all failed event either on error or negative result we get eventually the events in the correct order.
